sui_types/src/messages.rs

```diff
@@ -334,14 +335,65 @@ const_assert_eq!(
 
 */
 
-/// An transaction signed by a single authority
+// TODO: Should we have a uniform way to convert AuthoritySignedData to certificates?
+// Should we have a uniform authority certificate type too?
+
+// Generic types instantiated multiple times in the same Serde tracing session
+// requires a walk-around: https://crates.io/crates/serde-name
 #[derive(Debug, Eq, Clone, Serialize, Deserialize)]
-pub struct SignedTransaction {
-    pub transaction: Transaction,
+#[serde(remote = "AuthoritySignedData")]
+pub struct AuthoritySignedData<D> {
+    pub data: D,
     pub authority: AuthorityName,
     pub signature: AuthoritySignature,
 }
 
+impl<'de, T> Deserialize<'de> for AuthoritySignedData<T>
+where
+    T: Deserialize<'de>,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::de::Deserializer<'de>,
+    {
+        AuthoritySignedData::deserialize(DeserializeNameAdapter::new(
+            deserializer,
+            std::any::type_name::<Self>(),
+        ))
+    }
+}
+
+impl<T> Serialize for AuthoritySignedData<T>
+where
+    T: Serialize,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::ser::Serializer,
+    {
+        AuthoritySignedData::serialize(
+            self,
+            SerializeNameAdapter::new(serializer, std::any::type_name::<Self>()),
+        )
+    }
+}
+
+impl<D: Hash> Hash for AuthoritySignedData<D> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.data.hash(state);
+        self.authority.hash(state);
+    }
+}
+
+impl<D: Eq> PartialEq for AuthoritySignedData<D> {
+    fn eq(&self, other: &Self) -> bool {
+        self.data == other.data && self.authority == other.authority
```

Comment:
Can we add a comment here why we need this implemented and why we don't compare the signature too?